siteapp/views.py

```diff
@@ -25,30 +25,90 @@ def homepage(request):
     })
 
-def folder_list(request):
-    # Get the folders the user can see. Folders are accessible to a user
-    # just when they can see a Project within it or if they
-    # are an admin of the folder, so we go backwards from Projects
-    projects = Project.get_projects_with_read_priv(request.user, request.organization)
-    folders = (Folder.objects.filter(projects__in=projects)
-            | Folder.objects.filter(organization=request.organization, admin_users=request.user))\
-          .distinct()
 
-    # Count up the total number of projects in each folder. For admins, so they
-    # can know why they can't delete a folder.
-    from django.db.models import Count
-    folders = folders.annotate(project_count=Count('projects'))
+def assign_project_lifecycle_stage(projects):
+    # Define lifecycle stages.
+    # Because we alter this data structure in project_list,
+    # we need a new instance of it on every page load.
+    lifecycle_stages = [
+        {
+            "id": "none",
+            "label": "General Assessments",
+            "stage_col_width": { "xs-12" }, # "col_" + this => Bootstrap 3 column class
+            "stages": [
+                { "id": "none", "label": "", "subhead": "" },
+            ]
+        },
+        {
+            "id": "us_nist_rmf",
+            "label": "NIST Risk Management Framework",
+            "stage_col_width": { "md-2" }, # "col_" + this => Bootstrap 3 column class
+            "stages": [
+                { "id": "1_categorize", "label": "1. Categorize", "subhead": "Information System" },
+                { "id": "2_select", "label": "2. Select", "subhead": "Security Controls" },
+                { "id": "3_implement", "label": "3. Implement", "subhead": "Security Controls" },
+                { "id": "4_assess", "label": "4. Assess", "subhead": "Security Controls" },
+                { "id": "5_authorize", "label": "5. Authorize", "subhead": "Information System" },
+                { "id": "6_monitor", "label": "6. Monitor", "subhead": "Security Controls" },
+            ]
+        }
+    ]
+
+    # Create a mapping from concatenated lifecycle+stage IDs
+    # to tuples of (lifecycle object, stage object).
+    lifecycle_stage_code_mapping = { }
+    for lifecycle in lifecycle_stages:
+        for stage in lifecycle["stages"]:
+            lifecycle_stage_code_mapping[lifecycle["id"] + "_" + stage["id"]] = (
+                lifecycle,
+                stage
+            )
 
-    # Mark whether the user is an admin of any folders.
-    folders = list(folders)
-    for folder in folders:
-        folder.is_admin = (request.user in folder.get_admins())
+    # Load each project's lifecycle stage, which is computed by each project's
+    # root task's app's output document named govready_lifecycle_stage_code.
+    # That output document yields a string identifying a lifecycle stage.
+    for project in projects:
+        outputs = project.root_task.render_output_documents()
```

Comment:
Is calling `render_output_documents` going to force the top level app to render ALL output templates? Wouldn't that make determining the life cycle stage potentially very slow?

Reply:
The documents are actually lazy-rendered when they're accessed.

Reply:
i.e. Only the one document that we need for the life cycle stage stuff will actually be rendered.